hazelcast/src/main/java/com/hazelcast/map/impl/querycache/subscriber/AbstractInternalQueryCache.java

```diff
@@ -145,61 +152,125 @@ PartitioningStrategy getPartitioningStrategy() {
+        }
     }
 
-    protected void doFullValueScan(Predicate predicate, List<Data> resultingSet) {
-        InternalSerializationService serializationService = this.serializationService;
+    private boolean tryQueryOverIndexes(Predicate predicate, BiConsumer biConsumer) {
+        Iterable<QueryableEntry> query = indexes.query(predicate, SKIP_PARTITIONS_COUNT_CHECK);
+        if (query == null) {
+            return false;
+        }
 
-        CachedQueryEntry queryEntry = new CachedQueryEntry();
-        Set<Map.Entry<Data, QueryCacheRecord>> entries = recordStore.entrySet();
-        for (Map.Entry<Data, QueryCacheRecord> entry : entries) {
-            Data keyData = entry.getKey();
-            QueryCacheRecord record = entry.getValue();
-            Object value = record.getValue();
+        for (QueryableEntry entry : query) {
+            biConsumer.accept(entry.getKeyData(), entry.getValueData());
+        }
+        return true;
+    }
+
+    private void doFullScan(Predicate predicate, BiConsumer biConsumer) {
+        if (predicate == TruePredicate.INSTANCE) {
+            dumpAll(biConsumer);
+            return;
+        }
 
-            queryEntry.init(serializationService, keyData, value, extractors);
+        scanWithPredicate(predicate, biConsumer);
+    }
 
-            boolean valid = predicate.apply(queryEntry);
-            if (valid) {
-                resultingSet.add(queryEntry.getValueData());
+    /**
+     * Add all key-value pairs to result.
+     */
+    private void dumpAll(BiConsumer consumer) {
+        Set<Map.Entry<Object, QueryCacheRecord>> entries = recordStore.entrySet();
+        for (Map.Entry<Object, QueryCacheRecord> entry : entries) {
+            consumer.accept(entry.getKey(), entry.getValue().getRawValue());
+        }
+    }
+
+    /**
+     * Scan all key-value pairs and add matching ones with predicate to result.
+     */
+    private void scanWithPredicate(Predicate predicate, BiConsumer consumer) {
+        // needed for optimization where key and value are not an instance of Data type
+        final boolean areKeyValueObjectType = !queryCacheConfig.isSerializeKeys()
+                && InMemoryFormat.OBJECT == queryCacheConfig.getInMemoryFormat();
+
+        CachedQueryEntry queryEntry = new CachedQueryEntry(ss, extractors);
+        Set<Map.Entry<Object, QueryCacheRecord>> entries = recordStore.entrySet();
+        for (Map.Entry<Object, QueryCacheRecord> entry : entries) {
+            Object queryCacheKey = entry.getKey();
+            Object rawValue = entry.getValue().getRawValue();
+
+            if (areKeyValueObjectType) {
+                queryEntry.initWithObjectKeyValue(queryCacheKey, rawValue);
+            } else {
+                queryEntry.init(queryCacheKey, rawValue);
             }
+
+            if (!predicate.apply(queryEntry)) {
+                continue;
+            }
+
+            consumer.accept(queryCacheKey, queryEntry.getByPrioritizingObjectValue());
```

Comment:
Before line 273, there is a` predicate.apply` line and it can convert an internal Data type to object type while running. In this case, we can eliminate one unneeded deserialization by using already deserialized form of value with the help of `getByPrioritizingObjectValue`.